### ISSUE TITLE
OMC should reject impure function calls where not allowed

### ISSUE THREAD

**Issue by casella** [OpenModelica/OpenModelica#13036]
Consider this MWE
```modelica
model checkExist
  Boolean exists(start = false);
equation
  if (time > 0.0) then
    Modelica.Utilities.Streams.print(String(time),"file.txt");
  end if;
  exists = Modelica.Utilities.Files.exist("file.txt");
annotation(experiment(StartTime = -5, StopTime = 5, Tolerance = 1e-06, Interval = 0.02));
end checkExist;
```
The MLS [Sect. 12.3](https://specification.modelica.org/maint/3.6/functions.html#pure-modelica-functions) states:
> it is only allowed to call such a function from within:
> - Another function marked with the prefix impure.
> - A when-equation.
> - A when-statement.
> - pure(impureFunction(…)) – which allows calling impure functions in any pure context. The wrapping in pure(…) only by-passes the purity checking of the callee impureFunction; the argument expressions of the function call are not affected.
> - Initial equations and initial algorithms.
> - Binding equations for components declared as parameter – which is seen as syntactic sugar for having a parameter with fixed=false and the binding as an initial equation. 

The call in question obviously falls outside these categories, so the MWE should be rejected by the front-end.

**Comment by henrikt-ma** [comment]
Aren't you forgetting about this part?
> A deprecated semantics is that external functions (and functions defined in Modelica directly or indirectly calling them) without `pure` or `impure` keyword are assumed to be impure, but without any restriction on calling them. Except for the function `Modelica.Utilities.Streams.print`, a diagnostic must be given if called in a simulation model.

**Comment by maltelenz** [comment]
But it does have an `impure` keyword:

https://github.com/modelica/ModelicaStandardLibrary/blob/8e7876b5414475abab003c466ca49050812aeb01/Modelica/Utilities/Files.mo#L514

**Comment by henrikt-ma** [comment]
> But it does have an `impure` keyword:
> 
> https://github.com/modelica/ModelicaStandardLibrary/blob/8e7876b5414475abab003c466ca49050812aeb01/Modelica/Utilities/Files.mo#L514

Sorry, my bad; I didn't realize that this issue was about the `Modelica.Utilities.Files.exist` call, not the `Modelica.Utilities.Streams.print` call.

Regarding `Modelica.Utilities.Streams.print` I think one has to be careful with implementing too strict checking as long as the specification and the MSL are out of sync. I mean, the specification quite obviously has something in mind for how `Modelica.Utilities.Streams.print` should be handled, but the MSL does not make use of it.

**Comment by perost** [comment]
Implementing the rules in 12.3 would be fairly trivial, the frontend already tracks the information needed for it. But as mentioned the situation is a bit messy with regards to external functions that are not marked `pure` or `impure`.

If something is explicitly marked `pure` or `impure` it's probably safe to be strict with the rules, but to avoid breaking older code we might need to have an implicit and an explicit `impure` in the compiler so we know whether we should be strict or not. Currently we only track if something is `pure` or `impure`, but it's probably not a big deal to change.

**Comment by perost** [comment]
Looking into it a bit more we have another issue which is that we mark some builtin functions as impure as a way to say that they should never be evaluated, such as connect operators, `der`, `pre`, and many more. In some cases this is just unnecessary, e.g. `der` and `pre` will never be parameter expressions anyway. Some cases are a bit trickier, such as `Connection.isRoot` technically being evaluable but currently only handled at a specific point of the compilation.

**Comment by casella** [comment]
> Regarding `Modelica.Utilities.Streams.print` I think one has to be careful with implementing too strict checking as long as the specification and the MSL are out of sync. I mean, the specification quite obviously has something in mind for how `Modelica.Utilities.Streams.print` should be handled, but the MSL does not make use of it.

@henrikt-ma thanks for pointing this out! Would you mind opening a ticket on https://github.com/modelica/ModelicaStandardLibrary/issues so we can address this in 4.2.0?

**Comment by casella** [comment]
> I suspect that the idea behind making `Files.exist` `impure` is to prevent that it gets evaluated during translation

Sure. More in general, the file may be written or deleted during the simulation, possibly by the simulation itself, so even evaluating it only at runtime during initialization won't be a good idea.

> but being forced to mark it `impure` to get that effect comes at a very high price in terms of constraints on how the function may be used. Until the specification really offers something that makes sense to follow in libraries, it might be wise to not follow the specification too strictly…

OpenModelica tries to follow the specification as much as possible, given our development resources. If we implement something and we find out it's really not good, the idea is to apply feedback on MAP-Lang 😃

**Comment by henrikt-ma** [comment]
> @henrikt-ma thanks for pointing this out! Would you mind opening a ticket on https://github.com/modelica/ModelicaStandardLibrary/issues so we can address this in 4.2.0?

I'd rather not open an issue until I have some idea about what MAP-Lib should do, considering the current state of the specification.  I believe https://github.com/modelica/ModelicaSpecification/issues/3125 is a good introduction to the shortcoming of expressing what we need with only `pure`/`impure`.

**Comment by casella** [comment]
So you basically mean, pure/impure is still messy in the MLS, so the MSL should not bother for the time being?

**Comment by henrikt-ma** [comment]
> So you basically mean, pure/impure is still messy in the MLS, so the MSL should not bother for the time being?

I suppose it is good that the MSL at least tries to stay somewhat close to the specification, as it probably helps highlighting the problems so that they can be sorted out.  However, one must have realistic assumptions about how closely tools are able to follow the specification, and hence about how much tool acceptance actually says about conformance to the specification.

**Comment by casella** [comment]
@henrikt-ma I see you point. However, this ticket was opened in response to a [concrete use case](https://github.com/modelica/ModelicaStandardLibrary/issues/4472), where two tools give different results, because both fail to inform the modeller that the code is invalid. We should try to avoid this scenario as much as we can.

If the situation with impure is messy, maybe we should give warnings instead of rejecting the code outright. But we definitely should not accept it silently.

**Comment by henrikt-ma** [comment]
Yes, in situations like this we sometimes give funny messages of the kind:
> Warning: What you are doing here is illegal.
